pkg/rules/service.go

```diff
@@ -0,0 +1,197 @@
+	ctx, span := s.tracer.Start(ctx, "rules.Service.GetRule")
+	defer span.End()
+
+	rule, _, err := s.oathkeeper.GetRule(ctx, id).Execute()
+
+	if err != nil {
+		s.logger.Error(err.Error())
+		return nil, err
+	}
+
+	rules := make([]oathkeeper.Rule, 0)
+	rules = append(rules, *rule)
+	return rules, nil
+}
+
+func (s *Service) UpdateRule(ctx context.Context, id string, updatedRule oathkeeper.Rule) error {
+	ctx, span := s.tracer.Start(ctx, "rules.Service.UpdateRule")
+	defer span.End()
+
+	cm, err := s.k8s.ConfigMaps(s.cmNamespace).Get(ctx, s.cmName, metaV1.GetOptions{})
+
+	if err != nil {
+		s.logger.Error(err.Error())
+		return err
+	}
+
+	if _, ok := cm.Data[IdToFileName(id)]; !ok {
+		return fmt.Errorf("rule with ID %s not found", id)
+	}
+	if id != *updatedRule.Id {
+		return fmt.Errorf("The URL parameter id %s is different from payload rule id %s", id, *updatedRule.Id)
+	}
+
+	rawRule, err := json.Marshal(updatedRule)
+
+	if err != nil {
+		return err
+	}
+
+	cm.Data[IdToFileName(*updatedRule.Id)] = string(rawRule)
+
+	if _, err = s.k8s.ConfigMaps(s.cmNamespace).Update(ctx, cm, metaV1.UpdateOptions{}); err != nil {
+
+		return err
+	}
+
+	return nil
+}
+
+func (s *Service) CreateRule(ctx context.Context, newRule oathkeeper.Rule) error {
+	ctx, span := s.tracer.Start(ctx, "rules.Service.CreateRule")
+	defer span.End()
+
+	cm, err := s.k8s.ConfigMaps(s.cmNamespace).Get(ctx, s.cmName, metaV1.GetOptions{})
+
+	if err != nil {
+		s.logger.Error(err.Error())
+		return err
+	}
+
+	//Here I need to check the ID variable of all rules to make sure there's no collusion
+	for _, ruleRaw := range cm.Data {
+
+		extract := new(oathkeeper.Rule)
+
+		err := json.Unmarshal([]byte(ruleRaw), extract)
+
+		if err != nil {
+			s.logger.Errorf("failed unmarshalling %s - %v", ruleRaw, err)
+			return nil
+		}
+
+		if *newRule.Id == *extract.Id {
+			return fmt.Errorf("rule with ID %s already exists", *newRule.Id)
+		}
+	}
+
+	rawRule, err := json.Marshal(newRule)
+
+	if err != nil {
+		return err
+	}
+
+	cm.Data[IdToFileName(*newRule.Id)] = string(rawRule)
+
+	if _, err = s.k8s.ConfigMaps(s.cmNamespace).Update(ctx, cm, metaV1.UpdateOptions{}); err != nil {
+
+		return err
+	}
+
+	return nil
+}
+
+func (s *Service) DeleteRule(ctx context.Context, id string) error {
+	ctx, span := s.tracer.Start(ctx, "rules.Service.DeleteRule")
+	defer span.End()
+
+	cm, err := s.k8s.ConfigMaps(s.cmNamespace).Get(ctx, s.cmName, metaV1.GetOptions{})
+
+	if err != nil {
+		s.logger.Error(err.Error())
+		return err
+	}
+
+	if _, ok := cm.Data[IdToFileName(id)]; !ok {
+		return fmt.Errorf("rule with ID %s not found", id)
+	}
+
+	delete(cm.Data, IdToFileName(id))
+
+	if _, err = s.k8s.ConfigMaps(s.cmNamespace).Update(ctx, cm, metaV1.UpdateOptions{}); err != nil {
+
+		return err
+	}
+
+	return nil
+}
+
+func IdToFileName(id string) string {
+	filtered_name := regexp.MustCompile(`[^a-zA-Z0-9 ]+`).ReplaceAllString(id, "_")
```

Comment:
Add a comment explaining what this regex is doing, also do we want to allow the space (` `) character in the rule ID and the file name?

Let's be explicit about what we allow in the rule IDs, AFAICT with this approach if someone tried to create `rule-id-1` and `rule-id_1` it would cause an error even thought they have different rule IDs

Reply:
obsolete function